### DISCUSSION TITLE
A case which would benefit from dynamic shape compilation

### DISCUSSION THREAD

**Discussion by oliverdutton** [jax-ml/jax#16140] [Ideas]
I was looking at use cases popping [foldcomp](https://github.com/steineggerlab/foldcomp) on GPU by combining with [nerfax](https://github.com/PeptoneLtd/nerfax) but compilation times kills it from being competitive due to slightly different length arrays everywhere.

In meandering towards this I have code which does run on CPU, and has the potential to be competitive with the (in my opinion) more complex C++ code if compiled.

A colab showing the exact function and variation in shapes to compile is [here](https://github.com/PeptoneLtd/nerfax/blob/main/notebooks/jax_foldcomp_example.ipynb). 

I discussed this in https://github.com/iree-org/iree-jax/issues/57 where [iree](https://github.com/iree-org/iree-jax) seems like the project soonest to make it possible. @jpienaar @mattjj - I promised to share this as a general toy problem about 2 months ago (sorry for the ping-ing)

**Comment by jpienaar** [comment]
Hey, sorry I missed this while I was out and just noticed. Thanks for the example! @rsuderman we should dump this one a bit and see if can handle it

**Comment by oliverdutton** [comment]
Amazing, unfortunately one of the two dependencies, mdtraj, seems to fail installing by pip (even with older versions) as cython has changed, and pinning it doesn't seem to be working to fix it.

I'll track the [issue](https://github.com/mdtraj/mdtraj/issues/1800) and re-ping when the colab actually works

**Comment by oliverdutton** [comment]
The [colab example](https://github.com/PeptoneLtd/nerfax/blob/main/notebooks/jax_foldcomp_example.ipynb) now runs happily again (the cython version issues in the dependency is fixed)

As you can see
```
138 residues, 17.05 seconds with compilation, 0.036 Angstrom RMSD, 1.33 ms when compiled (with some compile time eval)
360 residues, 10.54 seconds with compilation, 0.030 Angstrom RMSD, 1.05 ms when compiled (with some compile time eval)
360 residues, 11.29 seconds with compilation, 0.030 Angstrom RMSD, 1.09 ms when compiled (with some compile time eval)
333 residues, 9.14 seconds with compilation, 0.033 Angstrom RMSD, 1.40 ms when compiled (with some compile time eval)
```
Compilation times are high and the program is run with many different shapes. Things could be padded, but then you pay the run time costs of excess computation and the time to do the padding itself.
